apps/evm_family/evm_sign_msg.c

```diff
@@ -0,0 +1,447 @@
+      !validate_initiate_query(&query->sign_msg.initiate) ||
+      !get_wallet_name_by_id(query->sign_msg.initiate.wallet_id,
+                             (uint8_t *)wallet_name,
+                             evm_send_error)) {
+    return false;
+  }
+
+  snprintf(msg,
+           sizeof(msg),
+           UI_TEXT_SIGN_PROMPT,
+           g_evm_app->lunit_name,
+           g_evm_app->name,
+           wallet_name);
+  // Take user consent to sign message for the wallet
+  if (!core_confirmation(msg, evm_send_error)) {
+    return false;
+  }
+
+  set_app_flow_status(EVM_SIGN_MSG_STATUS_CONFIRM);
+
+  memcpy(&(sign_msg_ctx.init),
+         &(query->sign_msg.initiate),
+         sizeof(query->sign_msg.initiate));
+
+  response.sign_msg.which_response = EVM_SIGN_MSG_RESPONSE_CONFIRMATION_TAG;
+  response.sign_msg.confirmation.dummy_field = 0;
+  evm_send_result(&response);
+
+  return true;
+}
+
+static bool get_msg_data(evm_query_t *query) {
+  bool result = false;
+  evm_result_t response = init_evm_result(EVM_RESULT_SIGN_MSG_TAG);
+  common_chunk_payload_t *chunk_payload = NULL;
+  uint32_t offset = 0;
+
+  sign_msg_ctx.msg_data = NULL;
+
+  do {
+    // Get next data chunk from host
+    if (!evm_get_query(query, EVM_QUERY_SIGN_MSG_TAG) ||
+        !check_which_request(query, EVM_SIGN_MSG_REQUEST_MSG_DATA_TAG) ||
+        false == query->sign_msg.msg_data.has_chunk_payload) {
+      break;
+    }
+
+    chunk_payload = &(query->sign_msg.msg_data.chunk_payload);
+
+    if (1 == chunk_payload->chunk_index) {
+      // After first chunk is received, allocate sufficient memory to msg_data
+      // member of @ref sign_msg_ctx
+      if (NULL == sign_msg_ctx.msg_data) {
+        sign_msg_ctx.msg_data = malloc(sign_msg_ctx.init.total_msg_size);
+        ASSERT(NULL != sign_msg_ctx.msg_data);
+      } else {
+        evm_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 0);
+        break;
+      }
+    }
+
+    // If chunk data is more than expected send error and break
+    if (sign_msg_ctx.init.total_msg_size <
+        chunk_payload->chunk.size + chunk_payload->remaining_size) {
+      evm_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                     ERROR_DATA_FLOW_INVALID_DATA);
+      break;
+    }
+
+    memcpy(sign_msg_ctx.msg_data + offset,
+           chunk_payload->chunk.bytes,
+           chunk_payload->chunk.size);
+    offset += chunk_payload->chunk.size;
+
+    if (sign_msg_ctx.init.total_msg_size == offset) {
+      result = true;
+    }
+
+    // Send chunk ack to host
+    response.sign_msg.which_response = EVM_SIGN_MSG_RESPONSE_DATA_ACCEPTED_TAG;
+    response.sign_msg.data_accepted.has_chunk_ack = true;
+    response.sign_msg.data_accepted.chunk_ack.chunk_index =
+        chunk_payload->chunk_index;
+    evm_send_result(&response);
+  } while (sign_msg_ctx.init.total_msg_size >= offset);
+
+  if (!result) {
+    free(sign_msg_ctx.msg_data);
+    sign_msg_ctx.msg_data = NULL;
+  }
```

Comment:
Free this at a common location

Reply:
updated as already handled on exit of caller